ingress/ingress.go

```diff
@@ -0,0 +1,150 @@
+	"k8s.io/client-go/kubernetes"
+	extentionslisters "k8s.io/client-go/listers/extensions/v1beta1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/cmd/kubeadm/app/util"
+
+	"github.com/argoproj/argo-rollouts/controller/metrics"
+	"github.com/argoproj/argo-rollouts/pkg/apis/rollouts/v1alpha1"
+	informers "github.com/argoproj/argo-rollouts/pkg/client/informers/externalversions/rollouts/v1alpha1"
+	controllerutil "github.com/argoproj/argo-rollouts/utils/controller"
+	ingressutil "github.com/argoproj/argo-rollouts/utils/ingress"
+	logutil "github.com/argoproj/argo-rollouts/utils/log"
+)
+
+const (
+	ingressIndexName = "byIngress"
+)
+
+type IngressController struct {
+	kubeclientset    kubernetes.Interface
+	rolloutsIndexer  cache.Indexer
+	rolloutSynced    cache.InformerSynced
+	ingressLister    extentionslisters.IngressLister
+	ingressSynced    cache.InformerSynced
+	rolloutWorkqueue workqueue.RateLimitingInterface
+	ingressWorkqueue workqueue.RateLimitingInterface
+	resyncPeriod     time.Duration
+
+	metricServer   *metrics.MetricsServer
+	enqueueRollout func(obj interface{})
+}
+
+// NewIngressController returns a new ingress controller
+func NewIngressController(
+	kubeclientset kubernetes.Interface,
+	ingressInformer extensionsinformers.IngressInformer,
+	rolloutsInformer informers.RolloutInformer,
+	resyncPeriod time.Duration,
+	rolloutWorkQueue workqueue.RateLimitingInterface,
+	ingressWorkQueue workqueue.RateLimitingInterface,
+	metricServer *metrics.MetricsServer) *IngressController {
+
+	controller := &IngressController{
+		kubeclientset:   kubeclientset,
+		rolloutsIndexer: rolloutsInformer.Informer().GetIndexer(),
+		rolloutSynced:   rolloutsInformer.Informer().HasSynced,
+		ingressLister:   ingressInformer.Lister(),
+		ingressSynced:   ingressInformer.Informer().HasSynced,
+
+		rolloutWorkqueue: rolloutWorkQueue,
+		ingressWorkqueue: ingressWorkQueue,
+		resyncPeriod:     resyncPeriod,
+		metricServer:     metricServer,
+	}
+
+	util.CheckErr(rolloutsInformer.Informer().AddIndexers(cache.Indexers{
+		ingressIndexName: func(obj interface{}) (strings []string, e error) {
+			if rollout, ok := obj.(*v1alpha1.Rollout); ok {
+				return ingressutil.GetRolloutIngressKeys(rollout), nil
+			}
+			return []string{}, nil
+		},
+	}))
+
+	ingressInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			controllerutil.Enqueue(obj, ingressWorkQueue)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			controllerutil.Enqueue(newObj, ingressWorkQueue)
+		},
+		DeleteFunc: func(obj interface{}) {
+			controllerutil.Enqueue(obj, ingressWorkQueue)
+		},
+	})
+	controller.enqueueRollout = func(obj interface{}) {
+		controllerutil.EnqueueRateLimited(obj, rolloutWorkQueue)
+	}
+
+	return controller
+}
+
+func (c *IngressController) Run(threadiness int, stopCh <-chan struct{}) error {
+	log.Info("Starting Ingress workers")
```

Comment:
would this be appropriate use for the logging context?

https://github.com/argoproj/argo-rollouts/blob/3b4b31da62893f333b8245d7a78eca8b9b00e071/utils/log/log.go#L23

Reply:
The IngressController runs outside of the context of a single Rollout. It's a whole different thread listening for changes to `Ingress` resources referenced in any of the observed Rollout resources.

Reply:
👍